Neos.Flow/Classes/Security/Cryptography/PrecomposedHashProvider.php

```diff
@@ -0,0 +1,60 @@
+<?php
+namespace Neos\Flow\Security\Cryptography;
+
+/*
+ * This file is part of the Neos.Flow package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Cache\Frontend\StringFrontend;
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Utility\Algorithms as UtilityAlgorithms;
+
+/**
+ * Precomposes a hash to be used to prevent timing attacks
+ *
+ * @Flow\Scope("singleton")
+ */
+class PrecomposedHashProvider
+{
+    /**
+     * @var HashService
+     * @Flow\Inject
+     */
+    protected $hashService;
+
+    /**
+     * @var StringFrontend
+     * @Flow\Inject
+     */
+    protected $cache;
+
+    /**
+     * @return float
+     */
+    public function getPrecomposedHash()
+    {
+        $hash = $this->cache->get('precomposed_hash');
+        if (!$hash) {
+            $hash = $this->precomposeHash();
```

Comment:
This would in theory also require the hashService to be non-lazily injected, but given this will at most run once on the initial request, which would take way longer than all other requests anyway, I think it's okay.